### PR TITLE
Define A Generic Lock Queue

### PR DESCRIPTION
**Pull request by chenhao94**

### CODE REVIEW

**Comment by xkszltl** [comment]
1. Why do you need a lock queue?
2. How about boost's lookfree queue?

File: includes/lock_queue.h [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+namespace cris::core {
+
+namespace impl {
+
+class LockQueueBase {
+   public:
+    using Self = LockQueueBase;
+
+    LockQueueBase(std::size_t capacity);
```
Comment (zhouxindong) [inline includes/lock_queue.h:17]: explicit?

**Comment by chenhao94** [comment]
> 1. Why do you need a lock queue?
> 2. How about boost's lookfree queue?

We need a thread safe message/job queue, and boost 's lockfree queue requires trivial dtor for the elements.

**Comment by xkszltl** [comment]
> > 1. Why do you need a lock queue?
> > 2. How about boost's lookfree queue?
> 
> We need a thread safe message/job queue, and boost 's lockfree queue requires trivial dtor for the elements.

Of course it needs trivial dtor. You shouldn't run dtor in critical section. If the job cannot clean up itself, use a lockfree gc queue for dtor.

**Comment by chenhao94** [comment]
> Of course it needs trivial dtor. You shouldn't run dtor in critical section. If the job cannot clean up itself, use a lockfree gc queue for dtor.

I don't understand how you should do it. Who owns the job then? After the job was sent and before it was picked up, who owns it? who takes care of the ref count if you somehow share it?

**Comment by xkszltl** [comment]
The actual job should not be shared, or you may break its semantic of single entry.
Ideally the job should clean up itself at the end of execution, but if it's not doable due to mem mgmt or whatever other reason, you can use another job queue dedicated for the clean up, or simply queue another job to the current queue.
Could you provide an example of how you want to use custom dtor?

**Comment by chenhao94** [comment]
We don't "share" the job. For example, we store `std::function<void()>` as "jobs" in the queue and workers fetch jobs from the queue. Nothing is shared. But those `std::function`s have non-trivial dtors, and we do need the resources being freed in the dtor. How are you going to gc them?

**Pull request closed without merging**